Mod Package Manager

sclu1034 opened the pull request:
This PR adds a new system to VMF that allows mod authors to dynamically load and unload resource packages in a fashion similar to the game's `PackageManager`.

The general process of loading a mod package is:

1. Acquire the package handle with `Mod.resource_package(mod_handle, package_name)`, where `package_name` is the usual path to a `.package` file and `mod_handle` is an ID defined by the engine's `stingray.Mod` interface.
2. Call `resource_package:load()` to start the loading process
3. Wait or call `resource_package:flush()`
4. Use the package's content
5. Call `resource_package:unload()` and `Mod.release_resource_package(resource_package)`.

The three major reasons why a central package manager is useful:

- the reference to `resource_package` has to be the same throughout the entire process
- there is no simple method call for a mod to it's `mod_handle`
- ease of use by adding guards and additional features such as asynchronous loading

The first point is dealt with by keeping package data in various tables based on the current state of the package.

~~To get access to a mod's `mod_handle` we need to iterate over the `Managers.mod._mods`. Since there is no directly shared property between those objects and VMF's mods, each mod, that wants to use the package manager will have to provide its own Workshop ID in its `mod_data`.~~
**Edit:** This is now handled for each mod during its call to `new_mod`.

~~To set up a small mod to test the package manager, see [these instructions](https://gist.github.com/SirAiedail/f32de0c53c835ebd7eef662be9016998).~~

sclu1034 committed:
Port mod package manager

sclu1034 committed:
Remove debug prints

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,135 @@
+  return false
+end
+
+function VMFMod:has_package_loaded(package_name)
```
Given that the methods `is_package_loading` and `has_package_loaded` check if a package is in different mutually exclusive states, both could be replaced by a single function that returns the current state of a package (eg,  `"not loaded"`, `"loading"` and `"loaded"`).

This would reduce the API surface area a bit. Thoughs?

sclu1034 replied:
I was trying to keep the API close to the game's `PackageManager`, partially because the mod I created this for (More Loading Screens) needs to be able to seemlessly choose between game packages and mod packages.

Apart from that, I would say, it's mostly personal taste if you like multiple booleans or a single state string.

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,135 @@
+function VMFMod:load_package(package_name, callback)
+  if not _initialized then
+    self:error("Package manager has not been initialized yet. It can only be used after the `all_mods_loaded` event has been processed.")
+    return
```
I find it a bit anti-intuitive that you can't start queueing asynchronously loaded packages as soon as a mod starts executing their main file.

Would it be feasible and worthwile to queue the calls and defer the processing until the `all_mods_loaded` event fires?

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,135 @@
+local _loading_package = nil
+local _loaded_packages = {}
+
+function VMFMod:load_package(package_name, callback)
```
This (and the other) functions have no argument type-checking. Ie,

```suggestion
function VMFMod:load_package(package_name, callback)
  if vmf.check_wrong_argument_type(self, "load_package", "package_name", package_name, "string") or
     vmf.check_wrong_argument_type(self, "load_package", "callback", callback, "function", "nil)
  then
    return
  end
```

Can you include these checks to keep the codebase consistent?

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,135 @@
+  local package_handle = string.format("resource_packages/%s/%s", self:get_name(), package_name)
+  local workshop_id = self:get_internal_data("workshop_id")
+
+  local resource_package = Mod.resource_package(_mod_handles[workshop_id], package_handle)
```
What would happen if the `package_handle` is invalid?
Also, I think an invalid `workshop_id` should be handled too.

Dreomite replied:
`workshop_id` shouldn't be passed via `mod_data` at all. It's always correct when you grab it from vanilla mod_manager.

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/vmf_mod_manager.lua:
```diff
@@ -115,6 +115,7 @@ function vmf.initialize_mod_data(mod, mod_data)
   vmf.set_internal_data(mod, "is_togglable",    mod_data.is_togglable or mod_data.is_mutator)
   vmf.set_internal_data(mod, "is_mutator",      mod_data.is_mutator)
   vmf.set_internal_data(mod, "allow_rehooking", mod_data.allow_rehooking)
+  vmf.set_internal_data(mod, "workshop_id",     mod_data.workshop_id)
```
I think we should try to automatically fill the `workshop_id` because:

+ It's more convenient for the user, and less error prone.
+ Makes it easier to fork mods of others.
+ Eliminates the need to provide two IDs for mods with beta/release versions.


The following patch (made with help from @Azumgi) should be able to determine the `workshop_id` for most mods, while still allowing it to be overriden for advanced use cases. _Warning: untested code & written from a browser._

```suggestion
local workshop_id = mod_data.workshop_id
local mod_manager = Manager.mods
local current_mod_load_index = mod_manager._mod_load_index
if not workshop_id and current_mod_load_index then
  workshop_id = mod_manager._mods[current_mod_load_index].id
end
vmf.set_internal_data(mod, "workshop_id", workshop_id)
```
(source: https://discordapp.com/channels/404978161726783489/454308427766693888/459282800793747457)

Dreomite replied:
As I told before `workshop_id` should not be passed in `mod_data` at all. There is just no point in that. Just straight up save ID from vanilla mod manager.

As for the piece of code in discord, I didn't test it. Even though, it should work.

Also, why mod_handle is not saved as internal data as well? It should be saved as internal data if there's no reason not to.

ManuelBlanc reviewed (commented):
Great work @SirAiedail, this looks pretty good!

I'm submitting some general feedback requesting some changes and discussing aspects.

Once this has bi's approval, it would be greatly appreciated if you could document the new module at 
[Vermintide-Mod-Framework/vmf-docs](https://github.com/Vermintide-Mod-Framework/vmf-docs)  so other modders can more easily find it.

Cheers.

Dreomite commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,135 @@
+    _loaded_packages[self] = {}
+  end
+
+  local package_handle = string.format("resource_packages/%s/%s", self:get_name(), package_name)
```
`resource_packages/mod_name/package_name` is just a convention and shouldn't be hardcoded. You should pass full path of the package to `mod:load_package` and not just its name.

Dreomite reviewed (commented):
I'm sorry I can't do a full review right now. Here are just a few things that do not require deep code analyzing. Will do a full review in a few days.

sclu1034 committed:
Automatically determine mod handle

Remove the burden fromt he modder to keep the workshop ID in sync by
determening the mod's handle during initial load.

This also removes the requirement to wait until all mods have been
loaded before being able to load packages.

sclu1034 committed:
Allow fuly qualified resource package names

Stop enforcing a naming convention for mod packages.

sclu1034 committed:
Check if package exists before attempting to load

ManuelBlanc committed:
Add parameter validation

Co-Authored-By: SirAiedail <SirAiedail@users.noreply.github.com>

sclu1034 committed:
Remove useless method

Azumgi commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,129 @@
+      _loaded_packages[loading_package.mod][loading_package.package_name] = loading_package.resource_package
+      _loading_package = nil
+
+      -- The callback has to be called last, so that any calls to `has_package_loaded` or `is_package_loading` return the correct value
```
This comment exceeds the maximum 120-char length. Split it in 2 lines.

Dreomite replied:
This comment exceeds maximum 120-char length. Split it into 2 lines.

Dreomite commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,129 @@
+
+  local resource_package = _loaded_packages[self][package_name]
+
+  resource_package:unload()
```
You can't use `unload` for mod packages.
https://help.autodesk.com/cloudhelp/ENU/Stingray-Help/lua_ref/ns_stingray_Mod.html#sig_stingray_Mod_resource__package

sclu1034 replied:
Can't find any documentation regarding `unload()` specifically. As I understand it, the `ResourcePackage` works like any other package, just with the added steps of `Mod.resource_package` to get it before `load:()` and `Mod.release_resource_package` to release it after `unload()`.

Dreomite commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,129 @@
+  local loading_package = _loading_package
+
+  if loading_package then
+    if loading_package.resource_package:has_loaded() then
```
Should add `else` and put `return` in here. Currently, when the package is loaded, it's going to call `return` anyway and skip loading of the next queued package until the next `update` call.

Dreomite commented on vmf/scripts/mods/vmf/modules/core/packages.lua:
```diff
@@ -0,0 +1,129 @@
+      _loading_package = nil
+
+      -- The callback has to be called last, so that any calls to `has_package_loaded` or `is_package_loading` return the correct value
+      loading_package.callback()
```
Execution of any external mod functions should be wrapped in safe call in order to not crash VMF in case of error.
```suggestion
      vmf.safe_call_nr(loading_package.mod, {"'%s' package loaded callback", loading_package.package_name},
                        loading_package.callback, loading_package.package_name)
```
I also added passing package name to the callback function so the modder can identify for which package callback was loaded. Is this a good idea? What do you think? If not, just remove the last argument.

Dreomite reviewed (commented):
Everything here is the subject for discussion.

1. Packages loaded with `VMFMod:load_package` should be considered async by default, even for mods without specified callback. Calling `flush` freezes the game until loading package is done and calling it inside `VMFMod:load_package` is a bad idea. Modders can use `VMFMod:has_package_loaded` for checking package status. You should specify the third boolean parameter telling explicitly that package is not async and `flush` should be called right away.

2. You should do `flush` for `_queued_packages` once they are loaded.

3. Add `vmf.unload_all_resource_packages` (or any other name) and call it from within `vmf_mod_object:on_reload` (defined in `vmf_loader.lua`) after `vmf.remove_custom_views()`. This function should go through all not unload packages and unload them. For every unloaded package, there should be warning with the package name, because it's modder's job to unload package properly.

4. `packages.lua` should be renamed to `vmf_package_manager.lua` to differentiate it from the vanilla `package_manager`. Just like `vmf_mod_loader` and `mod_loader`.

---

These are just style-related issues and I don't require you to do it yourself since there's still no style guide for this project. I can do it after merging.

1. All changed files were added with the new line at the end. I know, it's POSIX style, but in VMF all files end with the string.

2. Add two 120-char dividing comments for "VMFMod" and "VMF internal functions and variables".

3. Add 2 lines gap between function definitions.

4. Prefix `VMFMod` functions with their API description.

commands.lua is a nice example for (2), (3), and (4)

5) When defining a local variable and checking it in following `if` or `if not`, don't put an empty line between them.
```
local something = ...
if something then
 --[[...]]
end
```

Dreomite committed:
Wrap package loading callback in safe call

Co-Authored-By: SirAiedail <SirAiedail@users.noreply.github.com>

sclu1034 committed:
Add function documentation

sclu1034 committed:
Rework package loading timing

Loading packages asynchronously would unnecessarily skip a frame before
starting to load a new package after the previous one finished.

sclu1034 committed:
Flush packages after loading asynchronously

sclu1034 committed:
Make package loading asynchronous by default

sclu1034 committed:
Forcefully unload packages on VMF reload

sclu1034 committed:
Rename package manager module

sclu1034 committed:
Remove final new lines

sclu1034 committed:
Fix package loading timing

Azumgi commented:
Looking nice. Thanks!
